### PR TITLE
add github actions workflow and yaml syntax checks

### PR DESCRIPTION
**Pull request by matthias-heinisch**
this pull request implements automated syntax checks that are validating all contributions to this repository.

### CODE REVIEW

**Comment by uwe-salomon** [comment]
Example code for checking the YAML of a type spec:

~~~python
def pytest_generate_tests(metafunc):
    typeFiles = glob.glob('types/*')
    if "typeFile" in metafunc.fixturenames:
        metafunc.parametrize("typeFile", typeFiles)



def test_type_file(typeFile):
    assert re.match(r'types/csmim\.obj\..*\d+\.yaml$', typeFile) is not None
    
    fd = open(typeFile, 'r')
    ty = yaml.safe_load(fd)
    
    assert "id" in ty
    assert isinstance(ty["id"], str)
    assert len(ty["id"]) > 0
    assert re.match(r'^[a-z0-9_\-\.]+$', ty["id"]) is not None  # FIXME: Underscore not allowed?
    assert typeFile == "types/" + ty["id"] + ".yaml"
    
    assert "supertypes" in ty
    for supertype in ty["supertypes"]:
        assert len(glob.glob('types/' + supertype + '.yaml')) == 1
~~~

All but one check for the "id" attribute might be replaced by a schema check.

This example does multiple checks in one function, which we discussed is bad style. But it does the yaml conversion only once, which is desirable. Maybe the `pytest_generate_tests()` function can be extended to parametrize with a tuple that contains the file name and the yaml? Or maybe we can use a test fixture that loads the file once before executing the tests?

**Comment by uwe-salomon** [comment]
Example code for checking that there is no supertype cycle:

~~~python
class VisitSupertypes:
    def __init__(self, types):
        self.types   = types
        self.visited = set()
    
    def visit(self, ty):
        self.visited.add(ty["id"])
        for supertype in ty["supertypes"]:
            assert not supertype in self.visited
            self.visit(self.types[supertype])



def test_supertype_cycles():
    typeFiles = glob.glob('types/*')
    types     = {}
    
    for typeFile in typeFiles:
        try:
            fd = open(typeFile, 'r')
            ty = yaml.safe_load(fd)
            if typeFile == "types/" + ty["id"] + ".yaml":
                types[ty["id"]] = ty
        except:
            print("ignoring " + typeFile)  # FIXME: Insert placeholder in types dictionary?

    for tyId in types:
        VisitSupertypes(types).visit(types[tyId])
~~~

Again, if more tests are written that need the list of all types parsed into YAML, it would be great to do the expensive parsing (and ignoring errors caught by previous tests) only once.

**Comment by uwe-salomon** [comment]
The check_cycle() function looks good. I just noticed that it does not search for cycles but for "double supertypes", i.e. any case where a type file appears more than once when iterating through the supertypes. We did not specify in the standard whether that is legal and what it means, if I remember correctly.

In any case, I created a repository that collects test cases for the test suite, in order to verify it better: https://github.com/uwe-salomon/csmim-testsuite. See the `types/csmim.obj.circle.4.yaml` there for such a double-supertype case that is not a circle.

**Review by uwe-salomon — CHANGES_REQUESTED** [review]
Thanks for your work, especially for finding the schemas for yaml!

File: .github/workflows/csmim_schema.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+id: regex('^csmim\.obj\..[a-z0-9_\-\.]+\.\d+$')
+supertypes: list(required=False)
```
[the remark this answers is not in the document]
Comment (matthias-heinisch) [reply .github/workflows/csmim_schema.yaml]: well, I added the check with respect to the security concern in your other comment (regarding glob patterns)

File: .github/workflows/csmim_schema.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  mandatory: bool(required=False)
+  description: str()
+
+resource:
+  id: str()
+  mode: enum('r', 'w', 'x')
+  type: enum('void', 'enum', 'string', 'integer', 'bool')
+  mandatory: bool(required=False)
+  description: str()
+  enum-values: map(required=False)
```
[the remark this answers is not in the document]
Comment (matthias-heinisch) [reply .github/workflows/csmim_schema.yaml]: conditions in the schema are not supported and discouraged. We could introduce enum-resources and other-resources in the schema, but this gets complex quickly. So I will keep all possible options in the schema and add further tests with pytest to check this.

File: .github/workflows/csmim_schema.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  parameters: list(include('attribute'))
```
[the remark this answers is not in the document]
Comment (matthias-heinisch) [reply .github/workflows/csmim_schema.yaml]: conditions in the schema are not supported and discouraged. We could introduce enum-resources and other-resources in the schema, but this gets complex quickly. So I will keep all possible options in the schema and add further tests with pytest to check this.

File: .github/workflows/test-csmim-syntax.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        yamale.validate(csmim_schema, data)
+    except Exception as e:
+        print(e)
+        pytest.fail("Schema missmatch")
+
+
+# check the id of all type files matches the filename
+def test_id_matches(typeFile):
+    fd = open(typeFile, "r")
+    ty = yaml.safe_load(fd)
```
[the remark this answers is not in the document]
Comment (matthias-heinisch) [reply .github/workflows/test-csmim-syntax.py:35]: you are right, but it is just saving one line... If you want to do it, I support it.

File: .github/workflows/csmim_schema.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (uwe-salomon) [reply .github/workflows/csmim_schema.yaml]: I added such a test.

File: .github/workflows/csmim_schema.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (uwe-salomon) [reply .github/workflows/csmim_schema.yaml]: I added such a test.

**Comment by uwe-salomon** [comment]
I also removed the exception handling from the supertypes check to make the coder simpler. If an exception is raised, we can just fail the check with that exception. I tested it with an invalid supertypes spec.

**Review by uwe-salomon — APPROVED** [review]

**Comment by uwe-salomon** [comment]
Hey Matthias, I completed the "path/" tests and added the conflict tests for resource IDs and paths etc. I think we have a pretty good foundation now. The conflict tests are a bit complex, regrettably. What do you think?

**Review by uwe-salomon — APPROVED** [review]

File: .github/workflows/test-csmim-syntax.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+        assert os.path.exists(target)
+        target = os.path.normpath(target)
+        assert os.path.dirname(target) == "types"
+        assert os.path.basename(target) == os.path.basename(pathFile)
+
+
+# check that paths use only valid characters
+def test_path_dirs_valid(pathFile):
+    if os.path.isdir(pathFile):
+        assert re.fullmatch("[A-Za-z0-9\\-\\._]+", os.path.basename(pathFile))
```
Comment (matthias-heinisch) [inline .github/workflows/test-csmim-syntax.py:83]:
```suggestion
        assert re.fullmatch("[A-Za-z0-9\\-._]+", os.path.basename(pathFile))
```

**Review by matthias-heinisch — COMMENTED** [review]
looks great! only one tiny comment that I will adapt, then open for review.

**Review by uwe-salomon — APPROVED** [review]
Good work!

**Review by DennisWhetten-Boeing — APPROVED** [review]
I looked through the code and comments/discussion; I didn't check for efficacy or all functionality yet. From everything I've seen it looks great!

**Pull request merged**